code/controllers/subsystems/fluids.dm

```diff
@@ -57,137 +77,132 @@ var/datum/controller/subsystem/fluids/SSfluids
+				var/coming_from = GLOB.reverse_dir[spread_dir]
+				UPDATE_FLUID_BLOCKED_DIRS(neighbor_turf)
+				if((neighbor_turf.fluid_blocked_dirs & coming_from) || !neighbor_turf.CanFluidPass(coming_from) || checked[neighbor_turf])
 					continue
+				checked[neighbor_turf] = TRUE
+				var/obj/effect/fluid/other = locate() in neighbor_turf.contents
 				if(!other)
-					other = new /obj/effect/fluid(T)
-					other.temperature = F.temperature
-				F.equalizing_fluids += other
+					LOSE_FLUID(F, 1)
+					if(F.fluid_amount >= 1)
+						other = new /obj/effect/fluid(neighbor_turf)
+						SET_FLUID_DEPTH(other, 1)
+					else
+						qdel(F)
+						break
+		else
+			qdel(F)
 
 		if (MC_TICK_CHECK)
 			return
 
 	af_index = 1
-
 	while (af_index <= processing_fluids.len)
+
 		var/obj/effect/fluid/F = processing_fluids[af_index++]
 		if (QDELETED(F))
 			processing_fluids -= F
-		else
-			// Equalize across our neighbors. Hardcoded here for performance reasons.
-			if(!F.loc || F.loc != F.start_loc || !F.equalizing_fluids || !F.equalizing_fluids.len || F.fluid_amount <= FLUID_EVAPORATION_POINT)
-				continue
+			continue
 
-			F.equalize_avg_depth = 0
-			F.equalize_avg_temp = 0
-			F.flow_amount = 0
-
-			// Flow downward first, since gravity. TODO: add check for gravity.
-			if(F.equalizing_fluids.len >= downward_fluid_overlay_position)
-				var/obj/effect/fluid/downward_fluid = F.equalizing_fluids[downward_fluid_overlay_position]
-				if(downward_fluid.z == F.z-1) // It's below us.
-					F.equalizing_fluids -= downward_fluid
-					var/transfer_amount = min(F.fluid_amount, (FLUID_MAX_DEPTH-downward_fluid.fluid_amount))
-					if(transfer_amount > 0)
-						SET_FLUID_DEPTH(downward_fluid, downward_fluid.fluid_amount + transfer_amount)
-						LOSE_FLUID(F, transfer_amount)
-						if(F.fluid_amount <= FLUID_EVAPORATION_POINT)
-							continue
+		// Equalize across our neighbors. Hardcoded here for performance reasons.
+		if(!length(F.neighbors) || F.fluid_amount <= FLUID_EVAPORATION_POINT)
+			continue
 
-			var/setting_dir = 0
+		var/sufficient_delta = FALSE
+		var/equalize_avg_depth = 0
+		var/equalize_avg_temp = 0
+		F.last_flow_strength = 0
 
-			for(var/obj/effect/fluid/other in F.equalizing_fluids)
-				if(!istype(other) || QDELETED(other) || other.fluid_amount <= FLUID_DELETING)
-					F.equalizing_fluids -= other
-					continue
-				F.equalize_avg_depth += other.fluid_amount
-				F.equalize_avg_temp += other.temperature
+		for(var/thing in F.neighbors + F)
+			var/obj/effect/fluid/other = thing
+			if(abs(F.fluid_amount - other.fluid_amount) >= FLUID_EVAPORATION_POINT)
+				sufficient_delta = TRUE
+				break
 
+		if(sufficient_delta)
+			var/setting_dir = 0
+			var/equalize_count = 0
+			for(var/thing in F.neighbors + F)
+				var/obj/effect/fluid/other = thing
 				var/flow_amount = F.fluid_amount - other.fluid_amount
-				if(F.flow_amount < flow_amount && flow_amount >= FLUID_PUSH_THRESHOLD)
-					F.flow_amount = flow_amount
+				equalize_count++
```

Comment:
Feel like just do a length(F.neighbors) + 1 outside the loop.

Reply:
Not every neighbor equalizes.

Reply:
In that case it seems there's a logic issue, as this will be called exactly that many times here.